internal/gossiphttp/message.go

```diff
@@ -98,17 +150,16 @@ func writeMessage(w io.Writer, message []byte) error {
 	defer headerPool.Put(header)
 
 	if len(message) <= math.MaxUint16 {
-		header.data[0] = magic
-		binary.BigEndian.PutUint16(header.data[1:], uint16(len(message)))
+		header.setLength16(uint16(len(message)))
 	} else {
-		header.data[0] = magic32
-		binary.BigEndian.PutUint32(header.data[1:], uint32(len(message)))
+		header.setLength32(uint32(len(message)))
 	}
 
-	if _, err := w.Write(header.data); err != nil {
```

Comment:
I'm not sure all the other changes here are necessary, I think this could be fixed by truncating `header.data` on the call to Write:

```go
var headerSize int 

if len(message) <= math.MaxUint16 {
  header.data[0] = magic 
  binary.BigEndian.PutUint16(header.Data[1:], uint16(len(message)) 
  headerSize = 3 // or header16Size  
} else {
  header.data[0] = magic32 
  binary.BigEndian.PutUint32(header.Data[1:], uint32(len(message)) 
  headerSize = 5 // or header32Size  
}

if _, err := w.Write(header.data[:headerSize]); err != nil {
  return err 
}

_, err := w.Write(message)
return err 

```

Reply:
they are not necessary and your approach is clever and would work, but I find it easier to read and work with when the `header` struct knows best how to set the size correctly, seems like a more OOP approach

the read function would also become a bit more complex if we go with the approach of putting the logic inside the functions instead of `header` struct's methods

Reply:
maybe we can combine the both and I can use `header` struct's methods, but try to use the slice truncation and only one slice? let me know your thoughts

Reply:
I'm mainly concerned about the additional write and read calls. On a really busy cluster (like Loki's), one indirect call to the writer/reader interface isn't cheap. This is probably fine for small clusters (hundreds of nodes), but we were experiencing performance issues with a cluster at Loki's scale (multiple thousands of nodes).